// @iota/iota-sdk/dist/esm/keypairs/move-authenticator/signer.js
import { toBase64 } from "@iota/bcs";
import { bcs } from "../../bcs/index.js";
import { SIGNATURE_SCHEME_TO_FLAG, Signer } from "../../cryptography/index.js";
import { MoveAuthenticatorPublicKey } from "./publickey.js";
function getObjectIdFromCallArg(callArg) {
  if ("ImmOrOwnedObject" in callArg) {
    return callArg.ImmOrOwnedObject.objectId;
  } else if ("SharedObject" in callArg) {
    return callArg.SharedObject.objectId;
  } else {
    return callArg.Receiving.objectId;
  }
}
class MoveSigner extends Signer {
  /**
   * Creates a new MoveSigner with the provided MoveAuthenticator data.
   *
   * @param data - The MoveAuthenticator data containing object to authenticate, call arguments, and type arguments
   */
  constructor(data) {
    super();
    this.data = data;
  }
  /**
   * Get the key scheme for MoveAuthenticator
   */
  getKeyScheme() {
    return "MoveAuthenticator";
  }
  /**
   * Return the public key for this MoveAuthenticator.
   * Since MoveAuthenticator uses account abstraction, this returns a public key
   * based on the object ID.
   */
  getPublicKey() {
    const objectId = getObjectIdFromCallArg(this.data.objectToAuthenticate);
    return new MoveAuthenticatorPublicKey(objectId);
  }
  /**
   * Serialize the MoveAuthenticator data to bytes (without the signature scheme flag).
   * The data is already in the correct CallArg BCS format.
   *
   * @returns The BCS-serialized MoveAuthenticator bytes
   */
  async sign() {
    return bcs.MoveAuthenticator.serialize({
      callArgs: this.data.callArgs,
      typeArgs: this.data.typeArgs,
      objectToAuthenticate: {
        Object: this.data.objectToAuthenticate
      }
    }).toBytes();
  }
  /**
   * Override signWithIntent to handle MoveAuthenticator's special serialization format.
   * Unlike traditional signatures, MoveAuthenticator doesn't follow the standard
   * `flag || signature || publicKey` format.
   * It follows `flag || signature` format.
   */
  async signWithIntent(bytes, _intent) {
    const signature = await this.getSignature();
    return {
      signature,
      bytes: toBase64(bytes)
    };
  }
  /**
   * Generates the Move Authenticator signature.
   */
  async getSignature() {
    const serialized = await this.sign();
    const result = new Uint8Array(1 + serialized.length);
    result[0] = SIGNATURE_SCHEME_TO_FLAG[this.getKeyScheme()];
    result.set(serialized, 1);
    return toBase64(result);
  }
}
export {
  MoveSigner
};
