// @iota/iota-sdk/dist/cjs/keypairs/move-authenticator/builder.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var builder_exports = {};
__export(builder_exports, {
  InvalidMoveAuthAccountError: () => InvalidMoveAuthAccountError,
  InvalidMoveAuthArgError: () => InvalidMoveAuthArgError,
  MoveAuthenticatorBuilder: () => MoveAuthenticatorBuilder
});
module.exports = __toCommonJS(builder_exports);
class InvalidMoveAuthArgError extends Error {
  constructor(message) {
    super(message);
    this.name = "InvalidMoveAuthArgError";
  }
}
class InvalidMoveAuthAccountError extends Error {
  constructor(message) {
    super(message);
    this.name = "InvalidMoveAuthAccountError";
  }
}
class MoveAuthenticatorBuilder {
  /**
   * Create a new Move Authenticator builder from an object ID, which is the
   * sender of a transaction that this will be used to authenticate.
   *
   * @param objectToAuthenticate - The object ID of the objectToAuthenticate (sender of the transaction)
   */
  constructor(objectToAuthenticate) {
    this.callArgs = [];
    this.typeArgs = [];
    this.objectToAuthenticate = objectToAuthenticate;
  }
  /**
   * Set the move authenticator call inputs.
   * Call arguments must not be owned objects - they must be immutable or shared.
   *
   * @param args - Array of input kinds specifying the call arguments
   * @returns this builder for chaining
   */
  setCallArgs(args) {
    this.callArgs = args;
    return this;
  }
  /**
   * Add a single call argument.
   *
   * @param arg - The input kind to add
   * @returns this builder for chaining
   */
  addCallArg(arg) {
    this.callArgs.push(arg);
    return this;
  }
  /**
   * Add an immutable or owned object as a call argument.
   * Note: The object must actually be immutable when resolved.
   *
   * @param objectId - The object ID
   * @returns this builder for chaining
   */
  addImmutableObject(objectId) {
    this.callArgs.push({ ImmutableOrOwned: objectId });
    return this;
  }
  /**
   * Add a shared object as a call argument.
   *
   * @param objectId - The object ID
   * @param mutable - Whether the object is accessed mutably
   * @returns this builder for chaining
   */
  addSharedObject(objectId, mutable = false) {
    this.callArgs.push({ Shared: { objectId, mutable } });
    return this;
  }
  /**
   * Add a pure value as a call argument.
   *
   * @param value - The pure value bytes
   * @returns this builder for chaining
   */
  addPure(value) {
    this.callArgs.push({ Pure: value });
    return this;
  }
  /**
   * Set the move authenticator call type parameters.
   *
   * @param typeArgs - Array of type argument strings
   * @returns this builder for chaining
   */
  setTypeArgs(typeArgs) {
    this.typeArgs = typeArgs;
    return this;
  }
  /**
   * Add a single type argument.
   *
   * @param typeArg - The type argument string
   * @returns this builder for chaining
   */
  addTypeArg(typeArg) {
    this.typeArgs.push(typeArg);
    return this;
  }
  /**
   * Resolve this move authenticator builder into a MoveAuthenticator
   * which can be used to execute the given transaction.
   *
   * @param client - The IOTA client to use for fetching object data
   * @returns The resolved MoveAuthenticator data
   * @throws InvalidMoveAuthArgError if call arguments are invalid
   * @throws InvalidMoveAuthAccountError if the object is invalid
   */
  async finish(client) {
    const objectToAuthenticateResponse = await client.getObject({
      id: this.objectToAuthenticate,
      options: { showOwner: true }
    });
    if (!objectToAuthenticateResponse.data) {
      throw new InvalidMoveAuthArgError(
        `missing objectToAuthenticate ${this.objectToAuthenticate}`
      );
    }
    const objectToAuthenticateData = objectToAuthenticateResponse.data;
    const objectToAuthenticateOwner = objectToAuthenticateData.owner;
    const resolvedCallArgs = [];
    for (const input of this.callArgs) {
      if ("ImmutableOrOwned" in input) {
        const objectId = input.ImmutableOrOwned;
        const objResponse = await client.getObject({
          id: objectId,
          options: { showOwner: true }
        });
        if (!objResponse.data) {
          throw new InvalidMoveAuthArgError(`missing object ${objectId}`);
        }
        const objData = objResponse.data;
        const owner = objData.owner;
        if (owner !== "Immutable") {
          throw new InvalidMoveAuthArgError("call arguments must not be owned");
        }
        resolvedCallArgs.push({
          Object: {
            ImmOrOwnedObject: {
              objectId: objData.objectId,
              version: objData.version,
              digest: objData.digest
            }
          }
        });
      } else if ("Shared" in input) {
        const { objectId, mutable } = input.Shared;
        const objResponse = await client.getObject({
          id: objectId,
          options: { showOwner: true }
        });
        if (!objResponse.data) {
          throw new InvalidMoveAuthArgError(`missing object ${objectId}`);
        }
        const objData = objResponse.data;
        const owner = objData.owner;
        if (!owner || typeof owner !== "object" || !("Shared" in owner)) {
          throw new InvalidMoveAuthArgError(
            `object ${objectId} was passed as shared, but is not`
          );
        }
        resolvedCallArgs.push({
          Object: {
            SharedObject: {
              objectId,
              initialSharedVersion: owner.Shared.initial_shared_version,
              mutable
            }
          }
        });
      } else if ("Pure" in input) {
        resolvedCallArgs.push({
          Pure: {
            bytes: input.Pure
          }
        });
      }
    }
    let objectToAuthenticate;
    if (objectToAuthenticateOwner === "Immutable") {
      objectToAuthenticate = {
        ImmOrOwnedObject: {
          objectId: objectToAuthenticateData.objectId,
          version: objectToAuthenticateData.version,
          digest: objectToAuthenticateData.digest
        }
      };
    } else if (objectToAuthenticateOwner && typeof objectToAuthenticateOwner === "object" && "Shared" in objectToAuthenticateOwner) {
      objectToAuthenticate = {
        SharedObject: {
          objectId: objectToAuthenticateData.objectId,
          initialSharedVersion: objectToAuthenticateOwner.Shared.initial_shared_version,
          mutable: false
        }
      };
    } else {
      throw new InvalidMoveAuthAccountError(
        "objectToAuthenticate must be immutable or shared"
      );
    }
    return {
      callArgs: resolvedCallArgs,
      typeArgs: this.typeArgs,
      objectToAuthenticate
    };
  }
}
