// @iota/iota-sdk/dist/cjs/keypairs/move-authenticator/publickey.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var publickey_exports = {};
__export(publickey_exports, {
  MoveAuthenticatorPublicKey: () => MoveAuthenticatorPublicKey
});
module.exports = __toCommonJS(publickey_exports);
var import_bcs = require("@iota/bcs");
var import_publickey = require("../../cryptography/publickey.js");
var import_signature_scheme = require("../../cryptography/signature-scheme.js");
var import_iota_types = require("../../utils/iota-types.js");
var import_utils = require("@noble/hashes/utils");
class MoveAuthenticatorPublicKey extends import_publickey.PublicKey {
  /**
   * Creates a new MoveAuthenticatorPublicKey from an account object ID.
   *
   * @param authenticatedObjectId - The object ID as bytes (32 bytes)
   */
  constructor(authenticatedObjectId) {
    super();
    if (typeof authenticatedObjectId === "string") {
      this.authenticatedObjectId = (0, import_bcs.fromHex)(authenticatedObjectId);
    } else {
      this.authenticatedObjectId = authenticatedObjectId;
    }
  }
  /**
   * Return the byte array representation of the object ID
   */
  toRawBytes() {
    return this.authenticatedObjectId;
  }
  /**
   * Return the signature scheme flag for MoveAuthenticator
   */
  flag() {
    return import_signature_scheme.SIGNATURE_SCHEME_TO_FLAG.MoveAuthenticator;
  }
  /**
   * Return the IOTA address for this MoveAuthenticator.
   * Unlike other key types, the address IS the object ID directly,
   * not a hash of (flag || publicKey). This matches the Rust implementation:
   * `IotaAddress::from(object_id)`.
   */
  toIotaAddress() {
    return (0, import_iota_types.normalizeIotaAddress)((0, import_utils.bytesToHex)(this.authenticatedObjectId));
  }
  /**
   * Verification is not supported for MoveAuthenticator as it uses account abstraction.
   * The verification happens on-chain via the authenticator function.
   */
  async verify(_data, _signature) {
    throw new Error(
      "Verification is not supported for MoveAuthenticator. Verification happens on-chain via the authenticator function."
    );
  }
}
