// @iota/iota-sdk/dist/cjs/keypairs/move-authenticator/signer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var signer_exports = {};
__export(signer_exports, {
  MoveSigner: () => MoveSigner
});
module.exports = __toCommonJS(signer_exports);
var import_bcs = require("@iota/bcs");
var import_bcs2 = require("../../bcs/index.js");
var import_cryptography = require("../../cryptography/index.js");
var import_publickey = require("./publickey.js");
function getObjectIdFromCallArg(callArg) {
  if ("ImmOrOwnedObject" in callArg) {
    return callArg.ImmOrOwnedObject.objectId;
  } else if ("SharedObject" in callArg) {
    return callArg.SharedObject.objectId;
  } else {
    return callArg.Receiving.objectId;
  }
}
class MoveSigner extends import_cryptography.Signer {
  /**
   * Creates a new MoveSigner with the provided MoveAuthenticator data.
   *
   * @param data - The MoveAuthenticator data containing object to authenticate, call arguments, and type arguments
   */
  constructor(data) {
    super();
    this.data = data;
  }
  /**
   * Get the key scheme for MoveAuthenticator
   */
  getKeyScheme() {
    return "MoveAuthenticator";
  }
  /**
   * Return the public key for this MoveAuthenticator.
   * Since MoveAuthenticator uses account abstraction, this returns a public key
   * based on the object ID.
   */
  getPublicKey() {
    const objectId = getObjectIdFromCallArg(this.data.objectToAuthenticate);
    return new import_publickey.MoveAuthenticatorPublicKey(objectId);
  }
  /**
   * Serialize the MoveAuthenticator data to bytes (without the signature scheme flag).
   * The data is already in the correct CallArg BCS format.
   *
   * @returns The BCS-serialized MoveAuthenticator bytes
   */
  async sign() {
    return import_bcs2.bcs.MoveAuthenticator.serialize({
      callArgs: this.data.callArgs,
      typeArgs: this.data.typeArgs,
      objectToAuthenticate: {
        Object: this.data.objectToAuthenticate
      }
    }).toBytes();
  }
  /**
   * Override signWithIntent to handle MoveAuthenticator's special serialization format.
   * Unlike traditional signatures, MoveAuthenticator doesn't follow the standard
   * `flag || signature || publicKey` format.
   * It follows `flag || signature` format.
   */
  async signWithIntent(bytes, _intent) {
    const signature = await this.getSignature();
    return {
      signature,
      bytes: (0, import_bcs.toBase64)(bytes)
    };
  }
  /**
   * Generates the Move Authenticator signature.
   */
  async getSignature() {
    const serialized = await this.sign();
    const result = new Uint8Array(1 + serialized.length);
    result[0] = import_cryptography.SIGNATURE_SCHEME_TO_FLAG[this.getKeyScheme()];
    result.set(serialized, 1);
    return (0, import_bcs.toBase64)(result);
  }
}
